

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Preparation of the reference genome &mdash; xenopus-supplement 1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Assembly" href="star-assembly.html" />
    <link rel="prev" title="Cufflinks assembly of transcripts" href="reference.html" />
    <link href="../_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> xenopus-supplement
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../de-novo/de-novo.html">Trinity de novo assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genome-guided/genome-guided.html">Trinity genome-guided assembly</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="reference.html">STAR reference assembly</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="reference.html#gene-set-generation">Gene set generation</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Preparation of the reference genome</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#obtaining-the-necessary-files">Obtaining the necessary files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#run-star">Run STAR</a></li>
<li class="toctree-l4"><a class="reference internal" href="#convet-the-sam-files-so-that-they-are-ready-as-input-for-genome-assembly-and-feature-analysis">Convet the SAM files so that they are ready as input for genome assembly and feature analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#concatenate-the-bam-files">Concatenate the BAM files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#links">Links</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="star-assembly.html">Assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="ref-difexp.html">Differential expression analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../geneset-analysis.html">Gene set analysis</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../filtered-analysis.html">Filtered differential expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../geneset-analysis.html">Gene set analysis</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../qc/qc.html">Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../methods/database.html">Database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../gs/gene-sets.html">Functional gene sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../results/meta-data.html">Meta-analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">Works cited</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">xenopus-supplement</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="reference.html">Cufflinks assembly of transcripts</a> &raquo;</li>
        
      <li>Preparation of the reference genome</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/reference/reference-prep.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="preparation-of-the-reference-genome">
<h1>Preparation of the reference genome<a class="headerlink" href="#preparation-of-the-reference-genome" title="Permalink to this headline">¶</a></h1>
<p>In this section we download the reference genome for <em>X. tropicalis</em>, prepare the files, and map quality controlled reads to the genome.</p>
<div class="section" id="obtaining-the-necessary-files">
<h2>Obtaining the necessary files<a class="headerlink" href="#obtaining-the-necessary-files" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><p>obtain the genome and genome annotations</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ wget ftp://ftp.xenbase.org/pub/Genomics/JGI/Xentr7.1/xenopus_tropicalis_v7.1.tar.gz
~$ wget ftp://ftp.xenbase.org/pub/Genomics/JGI/Xentr7.1/Xentr7_2_Stable.gff3.gz
~$ tar xzf xenopus_tropicalis_v7.1.tar.gz
~$ tar xzf Xentr7_2_Stable.gff3.gz
</pre></div>
</div>
</li>
<li><p>convert gff3 to gtf file</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ gffread Xentr7_2_Stable.gff3 -T -o Xentr7_2_Stable.gtf
</pre></div>
</div>
</li>
<li><p>download xenbase gene mappings</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ wget ftp://ftp.xenbase.org/pub/GenePageReports/JgiToXenbaseGenePage_7.1.txt
</pre></div>
</div>
</li>
</ol>
</div>
<div class="section" id="run-star">
<h2>Run STAR<a class="headerlink" href="#run-star" title="Permalink to this headline">¶</a></h2>
<p>STAR is a powerful and fast program to align RNA-Seq data <a class="reference internal" href="../references.html#dobin13" id="id1"><span>[Dobin13]</span></a>.  A minimum of 30GB of RAM is recommended to run STAR.
Most of the following process is handled by a Python script and the generalized process is enumerated below.</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference download internal" download="" href="../_downloads/ef71aded83d2a907618031c6d09d3519/runSTAR.py"><code class="xref download docutils literal notranslate"><span class="pre">runSTAR.py</span></code></a></p></li>
</ul>
</div></blockquote>
<ol class="arabic">
<li><p>create a directory for the mapping and add necessary files</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ /usr/src/STAR/source/STAR --version
STAR_2.4.0j
~$ mkdir ~/sequencing/xenopus/star
~$ cp ./20100930/sequences/Xenopus_tropicalis.main_genome.scaffolds.fasta ~/sequencing/xenopus/star/genome.fa
~$ cp Xentr7_2_Stable.gtf ~/sequencing/xenopus/star/Xentr7_2_Stable.gtf
</pre></div>
</div>
</li>
<li><p>index the genome</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ /usr/src/STAR/source/STAR --runMode genomeGenerate --runThreadN <span class="m">24</span> --genomeDir ~/sequencing/xenopus/star
   --genomeFastaFiles ~/sequencing/xenopus/star/genome.fa --sjdbGTFfile ~/sequencing/xenopus/star/Xentr7_2_Stable.gtf
   --sjdbOverhang <span class="m">100</span>
</pre></div>
</div>
</li>
<li><p>align for each sample e.g.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ /usr/src/STAR/source/STAR --runThreadN <span class="m">24</span> --genomeDir --genomeDir ~/sequencing/xenopus/star
   --readFilesIn left_reads_sample_A.fq right_reads_sample_A.fq --outFileNamePrefix ~/sequencing/xenopus/star/A_
</pre></div>
</div>
</li>
</ol>
</div>
<div class="section" id="convet-the-sam-files-so-that-they-are-ready-as-input-for-genome-assembly-and-feature-analysis">
<h2>Convet the SAM files so that they are ready as input for genome assembly and feature analysis<a class="headerlink" href="#convet-the-sam-files-so-that-they-are-ready-as-input-for-genome-assembly-and-feature-analysis" title="Permalink to this headline">¶</a></h2>
<p>Each step need to be carried out on each sample. The following operations are carried out using SAMtools <a class="reference internal" href="../references.html#li09" id="id2"><span>[Li09]</span></a>.</p>
<ol class="arabic">
<li><p>convert the SAM to BAM file</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ <span class="nb">cd</span> ~/sequencing/xenopus/star
~$ /usr/bin/samtools view -b -S A_Aligned.out.sam &gt; A_aligned.bam
</pre></div>
</div>
</li>
<li><p>sort the BAM file</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$  /usr/bin/samtools sort -n A_aligned.bam A_aligned_sorted
</pre></div>
</div>
</li>
<li><p>convert the sorted BAM into SAM</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ /usr/bin/samtools view -h A_aligned_sorted.bam &gt; A_aligned_sorted.sam
</pre></div>
</div>
</li>
</ol>
<p>From the sorted SAM files we can obtain a count matrix for feature analysis.  The sorted BAM files are concatenated and used as input into the genome guided assembly.</p>
</div>
<div class="section" id="concatenate-the-bam-files">
<h2>Concatenate the BAM files<a class="headerlink" href="#concatenate-the-bam-files" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><p>merge the files</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ /usr/bin/samtools merge star_all_reads.bam A_aligned_sorted.bam B_aligned_sorted.bam ...
</pre></div>
</div>
</li>
<li><p>create a coordinate sorted bam file</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>~$ /usr/bin/samtools sort star_all_reads.bam star_all_reads_sorted.bam
</pre></div>
</div>
</li>
</ol>
</div>
<div class="section" id="links">
<h2>Links<a class="headerlink" href="#links" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://github.com/alexdobin/STAR">STAR on GitHub</a></p></li>
<li><p><a class="reference external" href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR Manual</a></p></li>
<li><p><a class="reference external" href="http://samtools.sourceforge.net/">SAMtools</a></p></li>
</ul>
</div></blockquote>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="star-assembly.html" class="btn btn-neutral float-right" title="Assembly" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="reference.html" class="btn btn-neutral float-left" title="Cufflinks assembly of transcripts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, AJ Richards et al.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>