

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Gene sets &mdash; xenopus-supplement 1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Meta-data analysis" href="../results/meta-data.html" />
    <link rel="prev" title="Database" href="../methods/database.html" />
    <link href="../_static/style.css" rel="stylesheet" type="text/css">

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> xenopus-supplement
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../de-novo/de-novo.html">Trinity de novo assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genome-guided/genome-guided.html">Trinity genome-guided assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/reference.html">STAR reference assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="../filtered-analysis.html">Filtered differential expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../geneset-analysis.html">Gene set analysis</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../qc/qc.html">Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../methods/database.html">Database</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Functional gene sets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#xenopus-annotation-summary">Xenopus annotation Summary</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#including-iea-evidence">Including IEA evidence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#excluding-iea-evidence">Excluding IEA evidence</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#annotation-of-other-model-organisms">Annotation of other model organisms</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">Including IEA evidence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Excluding IEA evidence</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#links">Links</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results/meta-data.html">Meta-analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">Works cited</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">xenopus-supplement</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Gene sets</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/gs/gene-sets.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="gene-sets">
<span id="gene-set-methods"></span><h1>Gene sets<a class="headerlink" href="#gene-sets" title="Permalink to this headline">¶</a></h1>
<p>We partition the composite genomes of multiple species into functional gene sets.  This is because <strong>Xenopus</strong> is not among the most completely annotated taxa.  In this analysis we consider several ways to generate gene sets.  More specifically, the annotation information used may come from a list of different species or we may include/exclude specific evidence codes.  The result is always a list of functional gene sets. The majority of annotations in the GO are automatically inferred and the are non-curated.  They use specifically the <code class="docutils literal notranslate"><span class="pre">IEA</span></code> evidence code.  When <code class="docutils literal notranslate"><span class="pre">IEA</span></code> is excluded we use only curated annotations.</p>
<p>We are most interested in the following two species of frog.  However, there are other well-annotated species that can help create functional gene sets.</p>
<blockquote>
<div><ul class="simple">
<li><p><em>Xenopus (Silurana) tropicalis</em> <a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=8364">[8364]</a></p></li>
<li><p><em>Xenopus laevis</em> <a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=8355">[8355]</a></p></li>
</ul>
</div></blockquote>
<table class="docutils align-center">
<colgroup>
<col style="width: 10%" />
<col style="width: 54%" />
<col style="width: 36%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Taxon ID</p></th>
<th class="head"><p>Scientific Name</p></th>
<th class="head"><p>Common Name</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>8355</p></td>
<td><p>Xenopus laevis</p></td>
<td><p>African clawed frog</p></td>
</tr>
<tr class="row-odd"><td><p>8364</p></td>
<td><p>Xenopus (Silurana) tropicalis</p></td>
<td><p>western clawed frog</p></td>
</tr>
</tbody>
</table>
<div class="section" id="xenopus-annotation-summary">
<h2>Xenopus annotation Summary<a class="headerlink" href="#xenopus-annotation-summary" title="Permalink to this headline">¶</a></h2>
<p>In general annotations may be associated directly with an NCBI GeneID or they may be associated with a gene product that has an assigned Uniprot accession (UniprotKB-AC) ID.  All UniprotKB-AC accessions are mapped to UniprotKb-entries.  To clarify the following tables.</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">taxon</span></code> - matches with the taxon summary table on the top of the page</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genes</span> <span class="pre">(annots)</span></code> - The number of unique GeneIDs from NCBI that are associated with a given taxon</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">uniprot</span> <span class="pre">(annots)</span></code>  - The identifiers from uniprot (non-unique) that are associated with a given taxon</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">coding-genes</span></code> - GeneIDs that have at least one associated uniprot identifier</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined</span></code> - uniprot and gene annotations with redundant annotations removed</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">coverage</span></code> - <code class="docutils literal notranslate"><span class="pre">combined</span></code> divided by <code class="docutils literal notranslate"><span class="pre">genes</span></code></p></li>
</ul>
</div></blockquote>
<p>The <code class="docutils literal notranslate"><span class="pre">combined</span></code> number of annotations may be less than the total observed annotations because multiple uniprot ids may map to a single gene id.</p>
<div class="section" id="including-iea-evidence">
<h3>Including IEA evidence<a class="headerlink" href="#including-iea-evidence" title="Permalink to this headline">¶</a></h3>
<table class="docutils align-center">
<colgroup>
<col style="width: 13%" />
<col style="width: 20%" />
<col style="width: 23%" />
<col style="width: 18%" />
<col style="width: 13%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Taxon</p></th>
<th class="head"><p>genes (annots)</p></th>
<th class="head"><p>uniprot (annots)</p></th>
<th class="head"><p>coding-genes</p></th>
<th class="head"><p>combined</p></th>
<th class="head"><p>coverage</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>8355</p></td>
<td><p>13531 (0)</p></td>
<td><p>3385 (3187)</p></td>
<td><p>2957</p></td>
<td><p>2773</p></td>
<td><p>20.4937</p></td>
</tr>
<tr class="row-odd"><td><p>8364</p></td>
<td><p>24817 (0)</p></td>
<td><p>1701 (1565)</p></td>
<td><p>1615</p></td>
<td><p>1485</p></td>
<td><p>5.9838</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="excluding-iea-evidence">
<h3>Excluding IEA evidence<a class="headerlink" href="#excluding-iea-evidence" title="Permalink to this headline">¶</a></h3>
<table class="docutils align-center">
<colgroup>
<col style="width: 13%" />
<col style="width: 20%" />
<col style="width: 23%" />
<col style="width: 18%" />
<col style="width: 13%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Taxon</p></th>
<th class="head"><p>genes (annots)</p></th>
<th class="head"><p>uniprot (annots)</p></th>
<th class="head"><p>coding-genes</p></th>
<th class="head"><p>combined</p></th>
<th class="head"><p>coverage</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>8355</p></td>
<td><p>13531 (0)</p></td>
<td><p>3385 (1254)</p></td>
<td><p>2957</p></td>
<td><p>1145</p></td>
<td><p>8.4621</p></td>
</tr>
<tr class="row-odd"><td><p>8364</p></td>
<td><p>24817 (0)</p></td>
<td><p>1701 (633)</p></td>
<td><p>1615</p></td>
<td><p>592</p></td>
<td><p>2.3855</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="annotation-of-other-model-organisms">
<h2>Annotation of other model organisms<a class="headerlink" href="#annotation-of-other-model-organisms" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-center">
<colgroup>
<col style="width: 8%" />
<col style="width: 58%" />
<col style="width: 34%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Taxon</p></th>
<th class="head"><p>Scientific Name</p></th>
<th class="head"><p>Common Name</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>10090</p></td>
<td><p>Mus musculus</p></td>
<td><p>house mouse</p></td>
</tr>
<tr class="row-odd"><td><p>9606</p></td>
<td><p>Homo sapiens</p></td>
<td><p>human</p></td>
</tr>
<tr class="row-even"><td><p>7227</p></td>
<td><p>Drosophila melanogaster</p></td>
<td><p>fruit fly</p></td>
</tr>
<tr class="row-odd"><td><p>7955</p></td>
<td><p>Danio rerio</p></td>
<td><p>leopard danio</p></td>
</tr>
</tbody>
</table>
<div class="section" id="id3">
<h3>Including IEA evidence<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<table class="docutils align-center">
<colgroup>
<col style="width: 13%" />
<col style="width: 20%" />
<col style="width: 23%" />
<col style="width: 18%" />
<col style="width: 13%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Taxon</p></th>
<th class="head"><p>genes (annots)</p></th>
<th class="head"><p>uniprot (annots)</p></th>
<th class="head"><p>coding-genes</p></th>
<th class="head"><p>combined</p></th>
<th class="head"><p>coverage</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>10090</p></td>
<td><p>69347 (18938)</p></td>
<td><p>16671 (15735)</p></td>
<td><p>15619</p></td>
<td><p>19006</p></td>
<td><p>27.4071</p></td>
</tr>
<tr class="row-odd"><td><p>9606</p></td>
<td><p>47736 (18114)</p></td>
<td><p>140289 (100435)</p></td>
<td><p>18921</p></td>
<td><p>18244</p></td>
<td><p>38.2185</p></td>
</tr>
<tr class="row-even"><td><p>7227</p></td>
<td><p>23387 (11563)</p></td>
<td><p>40464 (26100)</p></td>
<td><p>13543</p></td>
<td><p>11870</p></td>
<td><p>50.7547</p></td>
</tr>
<tr class="row-odd"><td><p>7955</p></td>
<td><p>36579 (14990)</p></td>
<td><p>2926 (2697)</p></td>
<td><p>2669</p></td>
<td><p>15069</p></td>
<td><p>41.1958</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="id4">
<h3>Excluding IEA evidence<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<table class="docutils align-center">
<colgroup>
<col style="width: 13%" />
<col style="width: 20%" />
<col style="width: 23%" />
<col style="width: 18%" />
<col style="width: 13%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Taxon</p></th>
<th class="head"><p>genes (annots)</p></th>
<th class="head"><p>uniprot (annots)</p></th>
<th class="head"><p>coding-genes</p></th>
<th class="head"><p>combined</p></th>
<th class="head"><p>coverage</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>10090</p></td>
<td><p>69347 (15763)</p></td>
<td><p>16671 (11428)</p></td>
<td><p>15619</p></td>
<td><p>15842</p></td>
<td><p>22.8445</p></td>
</tr>
<tr class="row-odd"><td><p>9606</p></td>
<td><p>47736 (14077)</p></td>
<td><p>140289 (17803)</p></td>
<td><p>18921</p></td>
<td><p>14263</p></td>
<td><p>29.8789</p></td>
</tr>
<tr class="row-even"><td><p>7227</p></td>
<td><p>23387 (10338)</p></td>
<td><p>40464 (12648)</p></td>
<td><p>13543</p></td>
<td><p>10500</p></td>
<td><p>44.8967</p></td>
</tr>
<tr class="row-odd"><td><p>7955</p></td>
<td><p>36579 (3962)</p></td>
<td><p>2926 (1261)</p></td>
<td><p>2669</p></td>
<td><p>4006</p></td>
<td><p>10.9516</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="links">
<h2>Links<a class="headerlink" href="#links" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="http://www-test.geneontology.org/page/current-go-statistics">Statistics about the current GO annotations</a></p></li>
<li><p><a class="reference internal" href="../methods/database.html"><span class="doc">Information about the database used for these data</span></a></p></li>
<li><p><a class="reference external" href="http://www.geneontology.org/page/guide-go-evidence-codes">Guide to GO evidence codes</a></p></li>
</ul>
</div></blockquote>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../results/meta-data.html" class="btn btn-neutral float-right" title="Meta-data analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../methods/database.html" class="btn btn-neutral float-left" title="Database" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, AJ Richards et al.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>